src/lib/pickles/inductive_rule.ml

```diff
@@ -1,150 +1,249 @@
+       , 'ret_value )
+       public_input =
+    | Input :
+        ('var, 'value) Impls.Step.Typ.t
+        -> ('var, 'value, 'var, 'value, unit, unit) public_input
+    | Output :
+        ('ret_var, 'ret_value) Impls.Step.Typ.t
+        -> ('ret_var, 'ret_value, unit, unit, 'ret_var, 'ret_value) public_input
+    | Input_and_output :
+        ('var, 'value) Impls.Step.Typ.t
+        * ('ret_var, 'ret_value) Impls.Step.Typ.t
+        -> ( 'var * 'ret_var
+           , 'value * 'ret_value
+           , 'var
+           , 'value
+           , 'ret_var
+           , 'ret_value )
+           public_input
+
+  (** The input type of an inductive rule's main function. *)
+  type 'public_input main_input =
+    { public_input : 'public_input
+          (** The publicly-exposed input to the circuit's main function. *)
+    }
+
+  (** The return type of an inductive rule's main function. *)
+  type ('prev_vars, 'widths, 'public_output, 'auxiliary_output) main_return =
+    { previous_proof_statements :
+        ('prev_vars, 'widths) H2.T(Previous_proof_statement).t
+    ; public_output : 'public_output
+    ; auxiliary_output : 'auxiliary_output
     }
 
-  module T
-      (Statement : T0)
-      (Statement_value : T0)
-      (Return_var : T0)
-      (Return_value : T0)
-      (Auxiliary_var : T0)
-      (Auxiliary_value : T0) =
+  module Make (M : sig
+    type _ t
+  end) =
   struct
-    type nonrec ('prev_vars, 'prev_values, 'widths, 'heights) t =
-      ( 'prev_vars
-      , 'prev_values
-      , 'widths
-      , 'heights
-      , Statement.t
-      , Statement_value.t
-      , Return_var.t
-      , Return_value.t
-      , Auxiliary_var.t
-      , Auxiliary_value.t )
-      t
+    type ( 'prev_vars
+         , 'prev_values
+         , 'widths
+         , 'heights
+         , 'a_var
+         , 'a_value
+         , 'ret_var
+         , 'ret_value
+         , 'auxiliary_var
+         , 'auxiliary_value )
+         t =
+      { identifier : string
+      ; prevs : ('prev_vars, 'prev_values, 'widths, 'heights) H4.T(Tag).t
+      ; main :
+             'a_var main_input
+          -> ('prev_vars, 'widths, 'ret_var, 'auxiliary_var) main_return M.t
+      ; feature_flags : bool Pickles_types.Plonk_types.Features.t
+      }
+
+    module T
+        (Statement : T0)
+        (Statement_value : T0)
+        (Return_var : T0)
+        (Return_value : T0)
+        (Auxiliary_var : T0)
+        (Auxiliary_value : T0) =
+    struct
+      type nonrec ('prev_vars, 'prev_values, 'widths, 'heights) t =
+        ( 'prev_vars
+        , 'prev_values
+        , 'widths
+        , 'heights
+        , Statement.t
+        , Statement_value.t
+        , Return_var.t
+        , Return_value.t
+        , Auxiliary_var.t
+        , Auxiliary_value.t )
+        t
+    end
   end
+
+  module Promise = Make (Promise)
+  module Deferred = Make (Async_kernel.Deferred)
+
+  (* This is the key fix - we need to include the Id-based module to match the MLI *)
```

Comment:
Can you explain a bit more this comment, @martyall?

Reply:
Ah, it seems to be used to introduce the phantom type and encapsulate the proof in it. Interesting.

Reply:
This was actually a not to self that should have been deleted. Will clean up when restoring comments in this file